code/modules/projectiles/pins.dm

```diff
@@ -0,0 +1,201 @@
+	var/emagged = FALSE
+	var/fail_message = "<span class='warning'>INVALID USER.</span>"
+	var/selfdestruct = 0 // Explode when user check is failed.
+	var/force_replace = 0 // Can forcefully replace other pins.
+	var/pin_removeable = 0 // Can be replaced by any pin.
+	var/obj/item/weapon/gun/gun
+
+
+
+/obj/item/device/firing_pin/New(newloc)
+	..()
+	if(istype(newloc, /obj/item/weapon/gun))
+		gun = newloc
+
+/obj/item/device/firing_pin/afterattack(atom/target, mob/user, proximity_flag)
+	if(proximity_flag)
+		if(istype(target, /obj/item/weapon/gun))
+			var/obj/item/weapon/gun/G = target
+			if(G.pin && (force_replace || G.pin.pin_removeable))
+				G.pin.forceMove(get_turf(G))
+				G.pin.gun_remove(user)
+				to_chat(user, "<span class ='notice'>You remove [G]'s old pin.</span>")
+
+			if(!G.pin)
+				gun_insert(user, G)
+				to_chat(user, "<span class ='notice'>You insert [src] into [G].</span>")
+			else
+				to_chat(user, "<span class ='notice'>This firearm already has a firing pin installed.</span>")
+
+/obj/item/device/firing_pin/emag_act(mob/user)
+	if(!emagged)
+		emagged = TRUE
+		to_chat(user, "<span class='notice'>You override the authentication mechanism.</span>")
+
+/obj/item/device/firing_pin/proc/gun_insert(mob/living/user, obj/item/weapon/gun/G)
+	gun = G
+	user.drop_from_inventory(src)
+	forceMove(gun)
+	gun.pin = src
+	return
+
+/obj/item/device/firing_pin/proc/gun_remove(mob/living/user)
+	gun.pin = null
+	gun = null
+	return
+
+/obj/item/device/firing_pin/proc/pin_auth(mob/living/user)
+	return 1
+
+/obj/item/device/firing_pin/proc/auth_fail(mob/living/user)
+	user.show_message(fail_message, 1)
+	if(selfdestruct)
+		user.show_message("<span class='danger'>SELF-DESTRUCTING...</span><br>", 1)
+		to_chat(user, "<span class='userdanger'>[gun] explodes!</span>")
+		explosion(get_turf(gun), -1, 0, 2, 3)
+		if(gun)
+			qdel(gun)
+
+/*
+Pins Below.
+*/
+
+//only used in wizard staffs/wands.
+/obj/item/device/firing_pin/magic
+	name = "magic crystal shard"
+	desc = "A small enchanted shard which allows magical weapons to fire."
+	icon_state = "firing_pin_wizwoz"
+
+// Test pin, works only near firing range.
+/obj/item/device/firing_pin/test_range
+	name = "test-range firing pin"
+	desc = "This safety firing pin allows weapons to be fired within proximity to a firing range."
+	fail_message = "<span class='warning'>TEST RANGE CHECK FAILED.</span>"
+	pin_removeable = 1
+	origin_tech = list(TECH_MATERIAL = 2, TECH_COMBAT = 2)
+
+/obj/item/device/firing_pin/test_range/pin_auth(mob/living/user)
+	var/area/A = get_area(src)
+	if (A && (A.flags & FIRING_RANGE))
+		return 1
+	else
+		return 0
+
+// Implant pin, checks for implant
+/obj/item/device/firing_pin/implant
+	name = "implant-keyed firing pin"
+	desc = "This is a security firing pin which only authorizes users who are implanted with a certain device."
+	fail_message = "<span class='warning'>IMPLANT CHECK FAILED.</span>"
+	var/obj/item/weapon/implant/req_implant = null
```

Comment:
This probably should just be `var/req_implant`, since the data being stored in this is a type, not an implant object.